kratos/containers/constraint_data.h

```diff
@@ -0,0 +1,324 @@
+*   c. equation ID
+*
+*   one should be able to access the constraint equation both with (dofID , dofKEY) pair and/or equationID of the slave.
+*   This equation is imposed on the linear system of equations either element wise or on the global system.
+*
+*/
+class ConstraintEquation
+{
+  private:
+    typedef Dof<double> DofType;
+    struct MasterData
+    {
+        KRATOS_CLASS_POINTER_DEFINITION(MasterData);
+        MasterData(DofType const &iMasterDof, double iWeight = 0.0) : masterWeight(iWeight), id(iMasterDof.Id()), key(iMasterDof.GetVariable().Key())
+        {
+        }
+        friend bool operator==(const MasterData::Pointer &iMasterDataOne, const DofType &iMasterDof)
+        {
+            return ((iMasterDataOne->MasterDofId() == iMasterDof.Id()) && (iMasterDataOne->MasterKey() == iMasterDof.GetVariable().Key()));
+        }
+        unsigned int MasterKey() { return key; }
+        double MasterWeight() { return masterWeight; }
+        void UpdateWeight(double iWeight) { masterWeight += iWeight; }
+        unsigned int MasterDofId() { return id; }
+        void SetEquationId(unsigned int iId) { equationId = iId; }
+        unsigned int MasterEqId() { return equationId; }
+        void SetMasterEqId(unsigned int iId) { equationId = iId; }
+
+      private:
+        double masterWeight;
+        const unsigned int id;
+        const unsigned int key;
+        unsigned int equationId;
+    };
+    typedef MasterData::Pointer MasterDataPointerType;
+    std::vector<MasterDataPointerType> mMasterDataVector;
+
+    const unsigned int id;
+    const unsigned int key;
+    unsigned int equationId;
+    double constant;
+    double constantUpdate;
+
+  public:
+    KRATOS_CLASS_POINTER_DEFINITION(ConstraintEquation);
+    typedef std::vector<MasterDataPointerType>::iterator iterator;
+    friend class ConstraintEquationContainer;
+
+    ConstraintEquation(DofType const &iSlaveDof) : id(iSlaveDof.Id()), key(iSlaveDof.GetVariable().Key())
+    {
+        SetConstant(0.0);
+        SetConstantUpdate(0.0);
+    }
+
+    void SetConstant(double iConstant) { constant = iConstant; }
+    void SetConstantUpdate(double iConstantUpdate) { constantUpdate = iConstantUpdate; }
+    double Constant() { return constant; }
+    double ConstantUpdate() { return constantUpdate; }
+    unsigned int SlaveDofId() { return id; }
+    unsigned int SlaveDofKey() { return key; }
+    unsigned int SlaveEquationId() { return equationId; }
+    void SetSlaveEquationId(unsigned int iId) { equationId = iId; }
+
+    // Add a master to this slave given are the masterDofId, masterDofKey, weight
+    void AddMasterData(DofType const &iMasterDof, double iWeight)
+    {
+        MasterDataPointerType masterData = MasterDataPointerType(new MasterData(iMasterDof, iWeight));
+        auto it = std::find(mMasterDataVector.begin(), mMasterDataVector.end(), iMasterDof);
+        if (it != mMasterDataVector.end()) // This master is already present so we add up the weight

```

Comment:
Yes. I update it if I find that the same master already exists. 
I put it together in the same function to avoid a second function like UpdateMasterData. This way from the outside its simple and one need not keep a track if they have already added a master.